reducer.py

```diff
@@ -0,0 +1,319 @@
+"""
+## Description
+
+We've been talking about custom reducers pretty seriously since mid-April.  Our
+major customers have asked for it and we should do it.  This is a proposal for
+how we could support them.
+
+I am hoping to have a discussion about this API at the next ml-ag meeting.
+
+**Conversation regarding validation metrics (#ml-ag, April 16):**
+    https://determined-ai.slack.com/archives/CSLAGUF3M/p1587050305213200
+
+**Conversation regarding training metrics (#ml-ag, April 17):**
+    https://determined-ai.slack.com/archives/CSLAGUF3M/p1587138289237300
+
+
+The API below addresses validation metrics, not training metrics, but it
+applies the following ideas proposed by @brain-good, @armandmcqueen , and
+@aaron276h during the discussion of training metrics:
+
+from @brain-good:
+
+>[we should] combine all the metrics and use a single function to aggregate the
+metrics.  It’s a lot easier to reason as a user.
+
+(see det.SimpleReducer, designed to meet this need)
+
+from @armandmcqueen :
+
+> Can we come up with a more user-friendly approach than concatenation - dict
+of lists/2d array?
+
+(see the det.Reducer, which doesn't do any magic and gives access full access
+to raw metrics)
+
+from @aaron276h:
+
+> If we keep [@brain-good's strategy vs a hierarchichal reduce] configurable,
+even if there is a noticable performance hits shouldn't be a big a problem
+
+(det.SimpleReducer as a subclass of det.Reducer is designed to meet this need)
+
+
+(I was supposed to do some profiling work to measure the performance overhead
+of @brain-good's strategy vs a hierarchichal reduce strategy, but the truth is
+that I have not yet had time, and the more I thought about it the more I like
+@aaron276h's idea of making it configurable)
+"""
+
+
+class Reducer:
+    """
+    A two-stage hierarchical reduction.  For a simpler interface, see SimpleReducer, below.
+
+    Unlike the current internal reducers, both steps will always run, and users who want a simpler
+    interface will let the first step be an identity function.
+    """
+
+    @abc.abstractmethod
+    def per_slot_reduce(self, metrics):
+        """
+        d-train optimization: do the bulk of your reduction on each GPU.
+
+        metrics is literally just a list of the metrics your trial returned, no collation at all.
+
+        This is always called, in single- and multi-gpu settings.
+        """
+        pass
+
+    @abc.abstractmethod
+    def cross_slot_reduce(self, per_slot_metrics):
+        """
+        This is always called, in single- and multi-gpu settings.
```

Comment:
Why does this step run in a single-GPU scenario? Is just for implementation consistency?

Reply:
If you didn't make this step always called, you would have to know in the first step if you were planning on calling the second step.  And example would be aggregating a confusion matrix to calculate mean IOU.  The per_slot_reduce() would do a sum, then the `cross_slot_reduce()` would do a sum then the mean IOU calculation.  If you didn't always call `cross_slot_reduce()` you would have to do the mean IOU calculation early.